Transformation

Addehman opened the pull request:
Created the script for Transforming the player character from werewolf to human.

Transform by holding "R"-key for 1.2 seconds which completes the transformation into human form. After 5 seconds it will return to the werewolf form again.

## Visualization
![transformation-show_00](https://user-images.githubusercontent.com/31672956/99541161-c3bf2900-29b0-11eb-854a-022778b0010f.gif)

Commit:
Transformation #7

Uploading the script for Transforming the player character from werewolf to human.

Transform by holding "R"-key for 1.2 seconds which completes the transformation into human form. After 5 seconds it will return to the werewolf form again.

MrJackey commented on Assets/Scripts/Transformation.cs:
```diff
@@ -0,0 +1,86 @@
+
+public class Transformation : MonoBehaviour
+{
+    //[SerializeField] private bool isTransformed = false;
```
Is this comment required to exist? Do we need it?

Addehman replied:
Ah nope!

MrJackey commented on Assets/Scripts/Transformation.cs:
```diff
@@ -0,0 +1,86 @@
+    private void Update() {
+		if (Input.GetKeyDown(KeyCode.R) && transStates == TransformationStates.Wolf) {
+            StartCoroutine(CoTransforming());
+			Debug.Log("Transformation process Charging...");
```
Do you think debug logging should be kept?

Addehman replied:
consider them gone!

MrJackey reviewed (changes requested):
Looking good, just a lot of debug which maybe should be removed

smeas commented on Assets/Scripts/Transformation.cs:
```diff
@@ -0,0 +1,86 @@
+
+public class Transformation : MonoBehaviour
+{
+    //[SerializeField] private bool isTransformed = false;
```
Brace should be on the same line.

Addehman replied:
ah ofc!

smeas commented on Assets/Scripts/Transformation.cs:
```diff
@@ -0,0 +1,86 @@
+
+
+	private IEnumerator CoTransforming() {
+    
```
Remove empty line at start of method.

Addehman replied:
done!

smeas commented on Assets/Scripts/Transformation.cs:
```diff
@@ -0,0 +1,86 @@
+	}
+
+
+	private IEnumerator HumanFormDuration() {
```
Use 'Co' prefix for coroutine?

Addehman replied:
ah! right! thanks!

smeas reviewed (changes requested):
Just some formatting/style issues.

There is mixed whitespace (spaces/tabs) in the whole file.

Addehman reviewed (commented):
Applied the changes! thanks for the reviews!

Commit:
Transformation - Changes

Applied changes from reviews.

Added the transformation "charge-up" when holding "r"-key - it turns grey. This is to signify that it's currently in the transformation process.

Commit:
Transformation

Converted Indentations to Tabs

MrJackey commented on Assets/Scripts/Transformation.cs:
```diff
@@ -25,44 +23,44 @@ public class Transformation : MonoBehaviour
 		}
 
-    	if (transStates == TransformationStates.Wolf) {
+    	if (transStates == TransformationStates.Transforming) {
```
Are you sure this should be .Transforming? Shouldn't this remain as .Wolf?

MrJackey commented on Assets/Scripts/Transformation.cs:
```diff
@@ -0,0 +1,83 @@
+
+
+	private IEnumerator CoTransforming() {
+		transStates = TransformationStates.Transforming;
```
What about having this as a boolean instead? Would be more generic if you could see if transforming and keep it from a middle. Might depend on if the human to wolf transformation should be instant.

MrJackey reviewed (changes requested):
Just one question, otherwise looks good

Commit:
Transformation - Changes

Removed TransformationState.Transforming - Added boolean isTransforming to do the same instead, which is to make the player grey while transforming from wolf to human.

Commit:
Transformation

Removed isTranforming boolean,
Made the code a bit ready for the future - for transforming to and back from human.

smeas commented on Assets/Scripts/Transformation.cs:
```diff
@@ -0,0 +1,91 @@
+
+
+	private IEnumerator CoTransforming() {
+		if (transStates == TransformationStates.Wolf) {
+
+			mySpriteRend.color = transformingColor;
+	
+			for (transTimer = transDuration; transTimer > 0 ; transTimer -= Time.deltaTime) {
+				if (Input.GetKeyUp(KeyCode.R)) {
+					mySpriteRend.color = wolfColor;
+					yield break;
+				}
+			yield return null;
+			}
+		}
+		else if (transStates == TransformationStates.Human) {
+
+			mySpriteRend.color = transformingColor;
+
+			for (transTimer = transDuration; transTimer > 0 ; transTimer -= Time.deltaTime) {
+				if (Input.GetKeyUp(KeyCode.R)) {
+					mySpriteRend.color = humanColor;
+					yield break;
+				}
+				yield return null;
+			}
+		}
```
Duplicate code in if statement blocks. The only difference is what color you set so you only need to check when you set the color.
```suggestion
		mySpriteRend.color = transformingColor;

		for (transTimer = transDuration; transTimer > 0 ; transTimer -= Time.deltaTime) {
			if (Input.GetKeyUp(KeyCode.R)) {
				if (transStates == TransformationStates.Wolf)
					mySpriteRend.color = wolfColor;
				else if (transStates == TransformationStates.Human)
					mySpriteRend.color = humanColor;

				yield break;
			}
			yield return null;
		}
```

smeas commented on Assets/Scripts/Transformation.cs:
```diff
@@ -0,0 +1,91 @@
+			TransformToHuman();
+			StartCoroutine(CoHumanFormDuration());
+		}
+
```
Remove empty line in the middle of if-else chain.

smeas reviewed (changes requested):
BUG: If you transform to wolf>human>wolf>human you get a shorter human form duration.
The problem is that the `CoHumanFormDuration` from the previous transformation is still running .

Commit:
Transformation

Changed the CoTransforming function to be more efficient - removed the duplicate if-statement.

Commit:
Transformation

added a stopper for the CoHumanFormDuration(), so that it won't make any unwanted transformations.

Commit:
Transform & Input Addition

Added the Transformation buttons into the Input Manager, both for Gamepad and keyboard - on the key R on keyboard, and button Y on gamepad.

Commit:
Transformation

Exposed the TransformationStates externally as ReadOnly.

MrJackey reviewed (approved):
+1
